pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java

```diff
@@ -805,29 +1033,49 @@ public TypeNode getElementTypeNode() {
     }
 
     @Override
-    public final VmList getTypeArgumentMirrors() {
+    public VmList getTypeArgumentMirrors() {
       return VmList.of(elementTypeNode.getMirror());
     }
 
     @Override
-    protected final PType doExport() {
+    protected PType doExport() {
       return new PType.Class(BaseModule.getListClass().export(), elementTypeNode.doExport());
     }
 
-    @Specialization
-    protected void eval(VirtualFrame frame, VmList value) {
-      if (skipElementTypeChecks) return;
+    @Override
+    public Object executeEagerly(VirtualFrame frame, Object value) {
+      if (!(value instanceof VmList vmList)) {
+        throw typeMismatch(value, BaseModule.getListClass());
+      }
+      if (skipElementTypeChecks) return vmList;
 
-      for (var elem : value) {
-        elementTypeNode.execute(frame, elem);
+      for (var elem : vmList) {
+        elementTypeNode.executeEagerly(frame, elem);
       }
 
-      LoopNode.reportLoopCount(this, value.getLength());
+      LoopNode.reportLoopCount(this, vmList.getLength());
+      return value;
     }
 
-    @Fallback
-    protected void fallback(Object value) {
-      throw typeMismatch(value, BaseModule.getListClass());
+    @Override
+    public Object execute(VirtualFrame frame, Object value) {
+      if (!(value instanceof VmList vmList)) {
```

Comment:
Is this a newly introduced scenario, or a previously missed one?

Reply:
This is neither; I rewrote this class to remove the specialization (see comment here: https://github.com/apple/pkl/pull/628/files#r1720334213)